app/models/manageiq/providers/azure/cloud_manager/event_catcher/stream.rb

```diff
@@ -1,67 +1,107 @@
+
+  # Creates an event monitor. Used internally by the Runner.
   #
   def initialize(ems)
     @ems = ems
     @collecting_events = false
     @since = nil
+    @connection = nil
   end
 
-  # Start capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should start/keep being captured.
+  #
   def start
     @collecting_events = true
   end
 
-  # Stop capturing events
+  # Sets a boolean used by the +each_batch+ method that indicates
+  # that events should stop being captured.
+  #
   def stop
     @collecting_events = false
   end
 
+  # Used internally by the Runner#monitor_events method.
+  #
   def each_batch
     while @collecting_events
-      yield get_events.collect { |e| JSON.parse(e) }
+      yield get_events.collect(&:to_hash)
     end
   end
 
   private
 
+  # Get a list of events that have happened since the most recent event time.
+  #
+  # Because Azure event timestamps are not necessarily stamped in order, an
+  # issue occurs where we could accidentally skip over events that happen
+  # in quick succession. We must therefore begin our query a couple minutes
+  # back from our most recent timestamp, and filter out any duplicates.
+  #
+  # See https://bugzilla.redhat.com/show_bug.cgi?id=1724312 for details.
+  #
   def get_events
-    # Grab only events for the last minute if this is the first poll
-    filter = @since ? "eventTimestamp ge #{@since}" : "eventTimestamp ge #{startup_interval}"
-    fields = 'authorization,description,eventName,eventTimestamp,resourceGroupName,resourceProviderName,resourceId,resourceType'
-    events = connection.list(:filter => filter, :select => fields, :all => true).sort_by(&:event_timestamp)
-
-    # HACK: the Azure Insights API does not support the 'gt' (greater than relational operator)
-    # therefore we have to poll from 1 millisecond past the timestamp of the last event to avoid
-    # gathering the same event more than once.
-    @since = one_ms_from_last_timestamp(events) unless events.empty?
+    filter = "eventTimestamp ge #{most_recent_time}"
+    events = connection.list(:filter => filter, :select => SELECT_FIELDS, :all => true)
+    self.since = events.max_by(&:event_timestamp)
     events
   end
 
+  # When the appliance first starts, or is restarted, start looking for events
+  # from a fixed, recent point in the past.
+  #
   def startup_interval
-    format_timestamp(1.minute.ago)
+    format_timestamp(2.minutes.ago)
   end
 
-  def one_ms_from_last_timestamp(events)
-    time = Time.at(one_ms_from_last_timestamp_as_f(events)).utc
-    format_timestamp(time)
-  end
-
-  def one_ms_from_last_timestamp_as_f(events)
-    Time.zone.parse(events.last.event_timestamp).to_f + 0.001
+  # Retrieve the most recent Azure event minus the timestamp buffer, or the
+  # startup interval if no records are found.
+  #
+  def most_recent_time
+    if since
+      format_timestamp(since - EVENT_TIMESTAMP_BUFFER)
+    else
+      startup_interval
```

Comment:
maybe just put this method contents here, to reduce amount of one-use methods?

Reply:
Was getting around to it. :)

Done.